File: connectors/go/ring.go
package impulsering

import (
	"time"
)

// Ring header field offsets (relative to base) — see spec/wire-format.md.
const (
	rMagic     = 0
	rCap       = 4
	rHead      = 64
	rProdLock  = 72
	rSpaceSeq  = 76
	rTail      = 128
	rDataSeq   = 136
	ringHeader = 192
	ringMagic  = 0x474E4952 // "RING"
	spinLimit  = 256
	maxWakeAll = 0x7FFFFFFF
)

func ringBytes(capacity int) int { return ringHeader + capacity }

type ring struct {
	data []byte
	base int
	cap  uint64
	mask uint64
}

func ringFormat(data []byte, base, capacity int) *ring {
	r := &ring{data: data, base: base, cap: uint64(capacity), mask: uint64(capacity) - 1}
	u64(data, base+rHead).Store(0)
	u64(data, base+rTail).Store(0)
	u32(data, base+rProdLock).Store(0)
	u32(data, base+rSpaceSeq).Store(0)
	u32(data, base+rDataSeq).Store(0)
	u32(data, base+rCap).Store(uint32(capacity))
	u32(data, base+rMagic).Store(ringMagic) // published last
	return r
}

func ringAttach(data []byte, base int) (*ring, error) {
	if u32(data, base+rMagic).Load() != ringMagic {
		return nil, errf("ring magic mismatch / not formatted")
	}
	capacity := u32(data, base+rCap).Load()
	if capacity == 0 || capacity&(capacity-1) != 0 {
		return nil, errf("ring capacity invalid")
	}
	return &ring{data: data, base: base, cap: uint64(capacity), mask: uint64(capacity) - 1}, nil
}

func (r *ring) dataOff() int { return r.base + ringHeader }

func (r *ring) writeWrapped(pos uint64, src []byte) {
	idx := int(pos & r.mask)
	off := r.dataOff()
	first := int(r.cap) - idx
	if first > len(src) {
		first = len(src)
	}
	copy(r.data[off+idx:], src[:first])
	if first < len(src) {
		copy(r.data[off:], src[first:])
	}
}

func (r *ring) readWrapped(pos uint64, n int) []byte {
	out := make([]byte, n)
	idx := int(pos & r.mask)
	off := r.dataOff()
	first := int(r.cap) - idx
	if first > n {
		first = n
	}
	copy(out, r.data[off+idx:off+idx+first])
	if first < n {
		copy(out[first:], r.data[off:off+(n-first)])
	}
	return out
}

// 3-state futex mutex on the producer side.
func (r *ring) lock() {
	lock := u32(r.data, r.base+rProdLock)
	if lock.CompareAndSwap(0, 1) {
		return
	}
	for lock.Swap(2) != 0 {
		futexWait(r.data, r.base+rProdLock, 2, 50)
	}
}

func (r *ring) unlock() {
	if u32(r.data, r.base+rProdLock).Swap(0) == 2 {
		futexWake(r.data, r.base+rProdLock, 1)
	}
}

// pushLocked returns 0 on success, 1 if full, -1 if the record can never fit.
func (r *ring) pushLocked(rec []byte) int {
	need := uint64(4 + len(rec))
	if need > r.cap {
		return -1
	}
	head := u64(r.data, r.base+rHead).Load()
	tail := u64(r.data, r.base+rTail).Load()
	if r.cap-(head-tail) < need {
		return 1
	}
	var lb [4]byte
	l := uint32(len(rec))
	lb[0], lb[1], lb[2], lb[3] = byte(l), byte(l>>8), byte(l>>16), byte(l>>24)
	r.writeWrapped(head, lb[:])
	r.writeWrapped(head+4, rec)
	u64(r.data, r.base+rHead).Store(head + need)
	return 0
}

func (r *ring) push(rec []byte, timeoutMs int) bool {
	start := time.Now()
	for {
		r.lock()
		rc := r.pushLocked(rec)
		if rc == 0 {
			r.unlock()
			u32(r.data, r.base+rDataSeq).Add(1)
			futexWake(r.data, r.base+rDataSeq, 1)
			return true
		}
		if rc < 0 {
			r.unlock()
			return false
		}
		seen := u32(r.data, r.base+rSpaceSeq).Load()
		r.unlock()
		if timeoutMs >= 0 {
			elapsed := int(time.Since(start).Milliseconds())
			if elapsed >= timeoutMs {
				return false
			}
			futexWait(r.data, r.base+rSpaceSeq, seen, timeoutMs-elapsed)
		} else {
			futexWait(r.data, r.base+rSpaceSeq, seen, 50)
		}
	}
}

func (r *ring) tryPop() []byte {
	tail := u64(r.data, r.base+rTail).Load()
	head := u64(r.data, r.base+rHead).Load()
	if head == tail {
		return nil
	}
	lb := r.readWrapped(tail, 4)
	n := int(lb[0]) | int(lb[1])<<8 | int(lb[2])<<16 | int(lb[3])<<24
	payload := r.readWrapped(tail+4, n)
	u64(r.data, r.base+rTail).Store(tail + 4 + uint64(n))
	u32(r.data, r.base+rSpaceSeq).Add(1)
	futexWake(r.data, r.base+rSpaceSeq, maxWakeAll)
	return payload
}

func (r *ring) popBlocking(timeoutMs int) []byte {
	start := time.Now()
	for {
		for i := 0; i < spinLimit; i++ {
			if v := r.tryPop(); v != nil {
				return v
			}
		}
		seen := u32(r.data, r.base+rDataSeq).Load()
		if v := r.tryPop(); v != nil {
			return v
		}
		if timeoutMs >= 0 {
			elapsed := int(time.Since(start).Milliseconds())
			if elapsed >= timeoutMs {
				return nil
			}
			futexWait(r.data, r.base+rDataSeq, seen, timeoutMs-elapsed)
		} else {
			futexWait(r.data, r.base+rDataSeq, seen, 100)
		}
	}
}

File: connectors/python/impulse_ring/ring.py
"""Shared-memory ring buffer (see SPEC/wire-format.md section 2).

Pure Python on top of the `_ring_native` atomics/futex helper. MPSC-safe via the
3-state futex mutex on the producer side; single consumer on the read side.
Length-prefixed records, wrap handled with a two-part copy.
"""

import time

from . import _ring_native as nat

# Field offsets (relative to ring base).
R_MAGIC = 0
R_CAP = 4
R_HEAD = 64
R_PRODLOCK = 72
R_SPACE_SEQ = 76
R_TAIL = 128
R_DATA_SEQ = 136
RING_HEADER = 192
RING_MAGIC = 0x474E4952  # "RING"
_U64 = (1 << 64) - 1
_INT_MAX = 0x7FFFFFFF
_SPIN = 256


def ring_bytes(capacity: int) -> int:
    return RING_HEADER + capacity


class Ring:
    def __init__(self, mm, base: int, cap: int):
        self.mm = mm
        self.base = base
        self.cap = cap
        self.mask = cap - 1
        self.data_off = base + RING_HEADER

    @classmethod
    def format(cls, mm, base: int, cap: int) -> "Ring":
        assert cap & (cap - 1) == 0, "capacity must be power of two"
        r = cls(mm, base, cap)
        nat.store_u64(mm, base + R_HEAD, 0)
        nat.store_u64(mm, base + R_TAIL, 0)
        nat.store_u32(mm, base + R_PRODLOCK, 0)
        nat.store_u32(mm, base + R_SPACE_SEQ, 0)
        nat.store_u32(mm, base + R_DATA_SEQ, 0)
        nat.store_u32(mm, base + R_CAP, cap)
        nat.store_u32(mm, base + R_MAGIC, RING_MAGIC)  # released last
        return r

    @classmethod
    def attach(cls, mm, base: int) -> "Ring":
        if nat.load_u32(mm, base + R_MAGIC) != RING_MAGIC:
            raise ValueError("ring magic mismatch / not formatted")
        cap = nat.load_u32(mm, base + R_CAP)
        if cap == 0 or (cap & (cap - 1)) != 0:
            raise ValueError("ring capacity invalid")
        return cls(mm, base, cap)

    # ---- data region copies (with wrap) ----
    def _write(self, pos: int, data: bytes):
        idx = pos & self.mask
        first = min(len(data), self.cap - idx)
        self.mm[self.data_off + idx : self.data_off + idx + first] = data[:first]
        if first < len(data):
            rest = len(data) - first
            self.mm[self.data_off : self.data_off + rest] = data[first:]

    def _read(self, pos: int, n: int) -> bytes:
        idx = pos & self.mask
        first = min(n, self.cap - idx)
        out = bytearray(self.mm[self.data_off + idx : self.data_off + idx + first])
        if first < n:
            out += self.mm[self.data_off : self.data_off + (n - first)]
        return bytes(out)

    # ---- producer 3-state futex mutex ----
    def _lock(self):
        if nat.cas_u32(self.mm, self.base + R_PRODLOCK, 0, 1):
            return
        while nat.swap_u32(self.mm, self.base + R_PRODLOCK, 2) != 0:
            nat.futex_wait(self.mm, self.base + R_PRODLOCK, 2, 50)

    def _unlock(self):
        if nat.swap_u32(self.mm, self.base + R_PRODLOCK, 0) == 2:
            nat.futex_wake(self.mm, self.base + R_PRODLOCK, 1)

    def _push_locked(self, data: bytes) -> int:
        need = 4 + len(data)
        if need > self.cap:
            return -1
        head = nat.load_u64(self.mm, self.base + R_HEAD)
        tail = nat.load_u64(self.mm, self.base + R_TAIL)
        used = (head - tail) & _U64
        if self.cap - used < need:
            return 1  # full
        self._write(head, len(data).to_bytes(4, "little"))
        self._write(head + 4, data)
        nat.store_u64(self.mm, self.base + R_HEAD, (head + need) & _U64)
        return 0

    def try_push(self, data: bytes) -> bool:
        self._lock()
        rc = self._push_locked(data)
        self._unlock()
        if rc == 0:
            nat.fetch_add_u32(self.mm, self.base + R_DATA_SEQ, 1)
            nat.futex_wake(self.mm, self.base + R_DATA_SEQ, 1)
            return True
        return False

    def push_blocking(self, data: bytes, timeout_ms: int = -1) -> bool:
        start = time.monotonic()
        while True:
            self._lock()
            rc = self._push_locked(data)
            if rc == 0:
                self._unlock()
                nat.fetch_add_u32(self.mm, self.base + R_DATA_SEQ, 1)
                nat.futex_wake(self.mm, self.base + R_DATA_SEQ, 1)
                return True
            if rc < 0:
                self._unlock()
                return False
            seen = nat.load_u32(self.mm, self.base + R_SPACE_SEQ)
            self._unlock()
            if timeout_ms >= 0:
                elapsed = int((time.monotonic() - start) * 1000)
                if elapsed >= timeout_ms:
                    return False
                nat.futex_wait(self.mm, self.base + R_SPACE_SEQ, seen, timeout_ms - elapsed)
            else:
                nat.futex_wait(self.mm, self.base + R_SPACE_SEQ, seen, 50)

    def try_pop(self):
        tail = nat.load_u64(self.mm, self.base + R_TAIL)
        head = nat.load_u64(self.mm, self.base + R_HEAD)
        if head == tail:
            return None
        length = int.from_bytes(self._read(tail, 4), "little")
        payload = self._read(tail + 4, length)
        nat.store_u64(self.mm, self.base + R_TAIL, (tail + 4 + length) & _U64)
        nat.fetch_add_u32(self.mm, self.base + R_SPACE_SEQ, 1)
        nat.futex_wake(self.mm, self.base + R_SPACE_SEQ, _INT_MAX)
        return payload

    def pop_blocking(self, timeout_ms: int = -1):
        start = time.monotonic()
        while True:
            for _ in range(_SPIN):
                v = self.try_pop()
                if v is not None:
                    return v
            seen = nat.load_u32(self.mm, self.base + R_DATA_SEQ)
            v = self.try_pop()
            if v is not None:
                return v
            if timeout_ms >= 0:
                elapsed = int((time.monotonic() - start) * 1000)
                if elapsed >= timeout_ms:
                    return None
                nat.futex_wait(self.mm, self.base + R_DATA_SEQ, seen, timeout_ms - elapsed)
            else:
                nat.futex_wait(self.mm, self.base + R_DATA_SEQ, seen, 100)
